mobile/test/java/org/chromium/net/CronetHttp3Test.java

```diff
@@ -83,57 +90,212 @@ public void log(int logLevel, String message) {
-    UrlRequest.Builder urlRequestBuilder2 =
-        cronvoyEngine.newUrlRequestBuilder(newUrl, callback2, callback2.getExecutor());
-    urlRequestBuilder2.build().start();
-    callback2.blockForDone();
+    TestUrlRequestCallback callback = new TestUrlRequestCallback();
+    ExperimentalUrlRequest.Builder urlRequestBuilder =
+        cronvoyEngine.newUrlRequestBuilder(testServerUrl, callback, callback.getExecutor());
+    urlRequestBuilder.addHeader("reset_after_request", "yes");
+    urlRequestBuilder.addHeader("content-type", "text");
+    urlRequestBuilder.setHttpMethod("POST");
+    TestUploadDataProvider dataProvider = new TestUploadDataProvider(
+        TestUploadDataProvider.SuccessCallbackMode.SYNC, callback.getExecutor());
+    dataProvider.addRead("test".getBytes());
+    urlRequestBuilder.setUploadDataProvider(dataProvider, callback.getExecutor());
+    // Set the request to be idempotent so Envoy knows it's safe to retry post-handshake
+    urlRequestBuilder.setIdempotency(ExperimentalUrlRequest.Builder.IDEMPOTENT);
 
-    // Verify the second request used HTTP/3
-    assertEquals(200, callback2.mResponseInfo.getHttpStatusCode());
-    assertEquals("h3", callback2.mResponseInfo.getNegotiatedProtocol());
+    urlRequestBuilder.build().start();
+    callback.blockForDone();
+
+    String stats = cronvoyEngine.getEnvoyEngine().dumpStats();
+
+    // Both HTTP/3 and HTTP/2 servers will reset after the request.
+    assertTrue(callback.mOnErrorCalled);
+    // Unlike testNoRetryPostPostHandshake there will be 3 requests - the initial HTTP/2 alt-svc
+    // request, the HTTP/3 request, and the HTTP/2 retry.
+    assertTrue(stats.contains("cluster.base.upstream_rq_total: 3"));
+    assertTrue(stats.contains("cluster.base.upstream_rq_retry: 1"));
+    // Because H/3 was disallowed on the final retry and TCP connected, H/3 gets marked as broken.
+    assertTrue(stats.contains("cluster.base.upstream_http3_broken: 1"));
+  }
+
+  @Test
+  @SmallTest
+  @Feature({"Cronet"})
+  public void testRetryPostHandshake() throws Exception {
+    setUp(printEnvoyLogs);
+
+    retryPostHandshake();
+  }
+
+  @Test
+  @SmallTest
+  @Feature({"Cronet"})
+  public void networkChangeAffectsBrokenness() throws Exception {
+    setUp(printEnvoyLogs);
+
+    // Set HTTP/3 to be marked as broken.
+    retryPostHandshake();
+
+    // From prior calls, there was one HTTP/3 connection established.
+    String pre_stats = cronvoyEngine.getEnvoyEngine().dumpStats();
+    assertTrue(pre_stats.contains("cluster.base.upstream_cx_http3_total: 1"));
+
+    // This should change QUIC brokenness to "failed recently".
+    cronvoyEngine.getEnvoyEngine().setPreferredNetwork(EnvoyNetworkType.WLAN);
+
+    // The next request may go out over HTTP/2 or HTTP/3 (depends on who wins the race)
+    // but HTTP/3 will be tried.
+    doBasicGetRequest();
+    String post_stats = cronvoyEngine.getEnvoyEngine().dumpStats();
```

Comment:
nit: `postStats`

Reply:
I wonder how hard it would be to add format checks for this... done.